### PR TITLE
add #define preprocess function module

### PR DESCRIPTION
**Pull request by ZJUnONE**

### CODE REVIEW

File: include/juicyc/preprocessor_impl.cc [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include "preprocessor_impl.h"
+#include <sstream>
+#include <string>
+#include <fstream>
+#include <map>
+#include <regex>
+using namespace std;
+namespace juicy {
+
+	void Preprocessor_impl::process_define()
```
Comment (tabokie) [inline include/juicyc/preprocessor_impl.cc]: Please take a look at google code style (I left a compact version in resource before). We should use CamelCase for class and function name in this project.

File: include/juicyc/preprocessor_impl.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		map <string, string> macro_map;
+		ifstream in;
+		string lineStr, segment, tmp = "", key, value;
+		for(auto it: input_files_names)
+		{
+			tmp = "";
+			in.open(it,ios::in);
+			if(!in)
+			{
+				cout << "File does not exist!" << endl;
```
Comment (tabokie) [inline include/juicyc/preprocessor_impl.cc]: Use juicyc::Status for error handling. e.g. return `Status::IOError("File does not exist")` here

File: include/juicyc/preprocessor_impl.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	char get();
+	std::string file_name();
+	uint32_t line_no();
+	uint32_t col_no();
+	//task scheduler
+	void push(std::string& file);
+	//added 
+	//noted, from opt -> input_files_names
+	//output_files -> opt
+	vector<string> input_files_names;
```
Comment (tabokie) [inline include/juicyc/preprocessor_impl.h]: Put data members in private or protected scope, and change it to xx_ format.

File: include/juicyc/preprocessor_impl.cc [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+				}
+			}
+			for(auto it: macro_map)
+			{
+				key = it.first;
+				value = it.second;
+				size_t length = tmp.size();
+				while((pos = tmp.find(key)) != string::npos)
+				{
+					regex re("[^_a-zA-Z0-9]*");
```
Comment (tabokie) [inline include/juicyc/preprocessor_impl.cc]: Note that C++ regex is extremely slow, we should avoid using it. This is an advancement advice, we can fix it in next PR.

File: include/juicyc/preprocessor_impl.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	vector<string> output_files;
+	Preprocessor_impl(CompilerOption& opt):
+		input_files_names(opt.files)
+	{}
+	~Preprocessor_impl() {}
+	void process_define();
```
Comment (tabokie) [inline include/juicyc/preprocessor_impl.h]: Preprocessor is not called by several pass. It should act like a stream filter only buffering limited characters. In this class, preprocessing routine is like:
```c++
bool get() {
    if(buffer_.empty()) {
        ReadBuffer();
        ProcessDefine(buffer_);
        ProcessInclude(buffer_);
    }
    return buffer_[cursor_];
}
```

File: include/juicyc/preprocessor_impl.cc (hunk 4 above)
Comment (ZJUnONE) [reply include/juicyc/preprocessor_impl.cc]: 这个单个字符的匹配，速度应该可以忽略不计吧。如果手动优化可能比用regex还要慢。

Comment (tabokie) [reply include/juicyc/preprocessor_impl.cc]: I presume regex is also slow at initialization, needs benchmark to prove. Here you only want to extract an identifier, why not find the first space instead.

**Comment by tabokie** [comment]
After last review, some remaining problems:
1. shouldn't use cout in a module. In normal case return Status is enough. If fatal error to happen, use `fprintf(stderr, "fatal error"); exit(1)` instead.
2. lots of format issues
3. Preprocessor adopt a stream-like protocol.
As a test PR, I will merge in and fix them myself.

**Pull request merged**